### PR TITLE
Service account storage configuration changes to support indexed SA watching

### PR DESCRIPTION
**Pull request by yt2985**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Currently we build a node agent which caches all the service accounts in the cluster using shared informer. This causes scalability issue as the total service accounts number in the cluster increases. We are trying to solve the scalability problem by only tracking service accounts used in the node. Inspired by the watch-based secret manager in the kubelet, we try to build a similar watch-based service account manager. As under each namespace, there is a default service account with the same `default` name. To avoid the scalability issue that may be caused by increasing number of namespace in the cluster, we need to use both the namespace and name as the index fields. 

This PR adds a matcher as the service account storage PredicateFunc and adds an AttrFunc and a TriggerFunc in StoreOptions. It also adds a unit test in the service account storage test and adds an integration test for the creating and watching service accounts. 

The matcher is implemented as:
         pkgstorage.SelectionPredicate {
		Label:       label,
		Field:       concat,
		GetAttrs:    GetAttrs,
		IndexFields: []string{"metadata.namespace/metadata.name"},
	}

The Label, Field and GetAttrs fields are originally default applied in the default SelectionPredicate in other layer. In this PR, we configure our own SelectionPredicate, where we added an IndexFields, which supports the O(1) watchers look-up in the cacher layer. 

Here the `concat` is the field selector that defaults as the input field selector. When the input field selector contains both selectors `metadata.namespace="xxx"` and `metadata.name="xxx"`, the `concat` will be added with a concatenated field selector that uses "metadata.namespace/metadata.name" as key. 

The TriggerFunc is used to trigger the creation of indexed watchers in the cacher layer.

func NamespaceNameTriggerFunc(obj runtime.Object) string {
	sa, ok := obj.(*api.ServiceAccount)
	if !ok {
		return ""
	}
	return fmt.Sprintf("%s/%s", sa.ObjectMeta.Namespace, sa.ObjectMeta.Name)
}

#### Which issue(s) this PR fixes:
This PR allows the kube-apiserver to support a watch-based service account manager without being overwhelmed by large amount of watching requests.

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
An `IndexFields: []string{"metadata.name"}` has been added to the service account storage SelectionPredicate. To trigger the use of indexed watchers for service account in the kube-apiserver, users can set a field selector, e.g. `filedSelector = fields.Set{"metadata.name": name}.AsSelector().String()` in the ListOptions of the client cache's ListWatch function.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None.
```

### CODE REVIEW

File: pkg/registry/core/serviceaccount/strategy.go [hunk 1]
```diff
@@ -79,6 +84,30 @@ func (strategy) WarningsOnUpdate(ctx context.Context, obj, old runtime.Object) [
...
+func Matcher(label labels.Selector, field fields.Selector) pkgstorage.SelectionPredicate {
+	return pkgstorage.SelectionPredicate{
+		Label:       label,
+		Field:       field,
+		GetAttrs:    GetAttrs,
+		IndexFields: []string{"metadata.name"},
+	}
+}
+
+// NameTriggerFunc returns value metadata.namespace of given object.
```
Comment (mikedanese) [inline pkg/registry/core/serviceaccount/strategy.go:106]: Comment looks wrong

File: pkg/registry/core/serviceaccount/storage/storage.go [hunk 2]
```diff
@@ -52,7 +54,11 @@ func NewREST(optsGetter generic.RESTOptionsGetter, issuer token.TokenGenerator,
 
 		TableConvertor: printerstorage.TableConvertor{TableGenerator: printers.NewTableGenerator().With(printersinternal.AddHandlers)},
 	}
-	options := &generic.StoreOptions{RESTOptions: optsGetter}
+	options := &generic.StoreOptions{
+		RESTOptions: optsGetter,
+		AttrFunc: serviceaccount.GetAttrs,
+		TriggerFunc: map[string]storage.IndexerFunc{"metadata.name": serviceaccount.NameTriggerFunc},
```
Comment (mikedanese) [inline pkg/registry/core/serviceaccount/storage/storage.go:60]: Shouldn't this be a namespaced name? This works for nodes because nodes aren't a namespaced resource, but service accounts will have name conflicts across namespaces.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by dims** [comment]
If you still need this PR then please remove the merge commits (see bot comment above), if not, please close the PR

**Comment by wojtek-t** [comment]
/assign

I would like to look at this PR as it's touching watchcache.

**Comment by yt2985** [comment]
Hi Wojciech,

Thank you very much!

Best wishes,

Ting Yu

**Comment by enj** [comment]
/hold

Reviewed during SIG Auth triage meeting.

Can we get an exact description (in the first message of this PR) of the problem this is trying to solve, and how that problem is specific only to service accounts, but in a way that funnels up all the way to the API server watch cache?

**Comment by yt2985** [comment]
Thank you, Mo Khan! I will update the description of the problem and
explain why we need the changes in the watch cache.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 3]
```diff
@@ -280,7 +280,7 @@ type Cacher struct {
 
 	// indexedTrigger is used for optimizing amount of watchers that needs to process
 	// an incoming event.
-	indexedTrigger *indexedTriggerFunc
+	indexedTrigger map[string]*indexedTriggerFunc
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:283]: If you're going to change this, please extract the changes done to cacher to a separate PR.
And you will for sure need a pretty extensive testing of that.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 4]
```diff
@@ -491,14 +487,35 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return nil, errors.NewServiceUnavailable(err.Error())
 	}
 
+	// Making triggerSupported be true requires the following conditions to be
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:490]: I think the usecase you're trying to address is basically "I want to watch individual SAs". In other words "watch per service-account".

For that you actually need an index on two fields (namespace, name), not two separate indexes on two fields.

The individual indexes are by definition independent, so to achieve (namespace, name) you effectively will be forced to to join the indexes, and we definitely don't want that.

So unless I misunderstood something, none of these changes aren't really needed.
All you need to do really is to define an index and trigger function to be sth like:
(metadata.namespace, metadata.name) and produce a value that is effectively some kind of concatenation of those two.

**Review by wojtek-t — COMMENTED** [review]
/hold

**Comment by wojtek-t** [comment]
This change doesn't look correct to me - PTAl at my comments in the code.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 4 above)
Comment (yt2985) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: Thank you very much, Wojciech. Following your comments, I rethought the whole logic and added an extra field selector which uses "metadata.namespace/metadata.name". Previously I was not familiar with the usage of field selector. Now I understand it.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 3 above)
Comment (yt2985) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: I removed the changes in cacher.go. Thank you!

File: pkg/registry/core/serviceaccount/strategy.go (hunk 1 above)
Comment (yt2985) [reply pkg/registry/core/serviceaccount/strategy.go]: Fixed.

File: pkg/registry/core/serviceaccount/storage/storage.go (hunk 2 above)
Comment (yt2985) [reply pkg/registry/core/serviceaccount/storage/storage.go]: Fixed.

**Comment by wojtek-t** [comment]
@enj - I want to make a second pass on it before this gets merged, so please wait with approval for me

**Comment by enj** [comment]
> One of the difference between secret and service account is: secret has global unique name across namespaces and service account only has unique name within the namespace. Thus for secret, we only need one index field (name) in the watch cache to filter out the secret and create the watcher. But for service account, we need both the namespace and name as the index fields.

I do not understand what this is saying.  Secrets and service accounts are both namespaced and thus their names are not globally unique.

What does kubelet do that is different from what you are trying to do?

**Comment by enj** [comment]
@yt2985 the scalability issue you appear to be running into seems to be caused by splitting one watch into N watches for a high volume resource.  It does not seem specific to service accounts from what I can tell.  Thus we should aim for a generic solution, and not one that arbitrarily targets your particular use case.

**Comment by yt2985** [comment]
@enj 
Thank you, Mo Khan. In kubelet, the watch-based manager uses a field selector `fieldSelector := fields.Set{"metadata.name": name}.AsSelector().String()` to invoke the secret watch cache: https://github.com/yt2985/kubernetes/blob/c34ddb5054e50a39e75cfc2eb7c5ca4a2145f6e6/pkg/kubelet/util/manager/watch_based_manager.go#L216
Previously Mike said we cannot use only `metadata.name` as the key to build the watch cache for service account. I may mis-understand what Mike said about the secret name. Now I recalled that Mike was actually saying that for service account, under each namespace, there is a default service account. Then if there are thousands of namespaces, there will be thousands of service accounts with the same `default` name. But for those service account tokens generated for these default service accounts, they will be the automatically added with a suffix and the name will be like "default-token-9rftj", which makes these secrets unique. So for secret, we can only use the `metadata.name` as the key, but for service account, we need both the `metadata.name` and  `metadata.namespace`.

**Comment by liggitt** [comment]
> In kubelet, the watch-based manager uses a field selector fieldSelector := fields.Set{"metadata.name": name}.AsSelector().String() to invoke the secret watch cache

that is also scoping the watch to a particular namespace, so the combined filter is both namespace and name. Nothing about the secret watch relies on the name being unique across namespaces

> for service account, we need both the metadata.name and metadata.namespace

I don't understand why a particular client watching objects within namespaces requires server changes

**Comment by enj** [comment]
@yt2985 based on your description, this code change is not correct.  When you issue a watch for a single, namespaced item, the namespace is part of the path, i.e. `kubectl get serviceaccount -n my-namespace --field-selector=metadata.name=default -w` results in the following path and query: `/api/v1/namespaces/my-namespace/serviceaccounts?fieldSelector=metadata.name=default&watch=true`.  This will only ever return updates for the `default` service account in the namespace `my-namespace`.  This single item watch was optimized with an index for secrets in #64257. @wojtek-t how expensive are these indexes?  Could we just change the defaults in `CompleteWithOptions` to have this `metadata.name` index for all resources?

**Comment by yt2985** [comment]
Here the triggerValue is produced by the index field value. https://github.com/kubernetes/kubernetes/blob/cf14b50b0d2215f039a6c1109445f01a7a222c21/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L499

Then the triggerValue is used as a key in the indexed watcher map: https://github.com/kubernetes/kubernetes/blob/cf14b50b0d2215f039a6c1109445f01a7a222c21/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L552

If the triggerValue is only the service account name, then there will be thousands of watches for the `default` service account in thousands of namespaces added to the same watches array using the same `default` key. https://github.com/kubernetes/kubernetes/blob/cf14b50b0d2215f039a6c1109445f01a7a222c21/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L134

Then the look up for the watch of `default` service account in a specific namespace in the cacher layer is not O(1), but O(n).

**Comment by liggitt** [comment]
Not partitioning namespace-scoped watchers seems like a general issue in the watch cache that should be fixed, not something to work around with a change to serviceaccounts

**Comment by liggitt** [comment]
I took a stab at a quick sketch of partitioning watchers by their namespace/name scope in https://github.com/kubernetes/kubernetes/pull/115532 in case that's interesting

**Comment by wojtek-t** [comment]
I spend some time thinking about Jordan's suggestion and I agree it's much better - see my comments in https://github.com/kubernetes/kubernetes/pull/115532/files#r1098568668

@yt2985 - will you be able to pick up that PR and ensure it's properly tested?

**Comment by yt2985** [comment]
Hi @wojtek-t , I am happy to pick up Jordan's PR and add full test coverage.

**Comment by wojtek-t** [comment]
> Hi @wojtek-t , I am happy to pick up Jordan's PR and add full test coverage.

@yt2985 - great!
Please make them generic enough by putting them here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go
so that we can actually run the same set of tests against all implementations of storage interface.

**Comment by wojtek-t** [comment]
We're taking a different approach to solve that problem - https://github.com/kubernetes/kubernetes/pull/115918 is close to be ready for merging.
So I'm closing this one in favor of the other one.
